src/main/cpp/checkpoint/CheckPoint.cpp

```diff
@@ -0,0 +1,675 @@
+{
+	std::stringstream ss;
+	ss << "Simulation " << groupnum;
+	std::string groupname = ss.str();
+
+	htri_t exist = H5Lexists(m_file, groupname.c_str(), H5P_DEFAULT);
+	if (exist <= 0) {
+		FATAL_ERROR("Non-existing group");
+	}
+
+	hid_t group = H5Gopen2(m_file, groupname.c_str(), H5P_DEFAULT);
+
+	hid_t f = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+
+	auto op_func = [&f](hid_t loc_id, const char* name, const H5L_info_t* info, void* operator_data) {
+		H5Ocopy(loc_id, name, f, name, H5P_DEFAULT, H5P_DEFAULT);
+		return 0;
+	};
+	auto temp = [](hid_t loc_id, const char* name, const H5L_info_t* info, void* operator_data) {
+		(*static_cast<decltype(op_func)*>(operator_data))(loc_id, name, info, operator_data);
+		return 0;
+	};
+
+	H5Literate(group, H5_INDEX_NAME, H5_ITER_NATIVE, nullptr, temp, &op_func);
+
+	H5Fclose(f);
+	H5Gclose(group);
+}
+
+Calendar CheckPoint::LoadCalendar(boost::gregorian::date date)
+{
+	htri_t exist = H5Lexists(m_file, "Config", H5P_DEFAULT);
+	if (exist <= 0) {
+		FATAL_ERROR("Invalid file");
+	}
+
+	WriteDSetFile("tmp_holidays.json", "Config/holidays");
+
+	Calendar result;
+	result.Initialize(date, "tmp_holidays.json");
+	boost::filesystem::remove("tmp_holidays.json");
+	return result;
+}
+
+void CheckPoint::WriteClusters(const ClusterStruct& clusters, boost::gregorian::date date)
+{
+
+	std::string datestr = to_iso_string(date);
+	htri_t exist = H5Lexists(m_file, datestr.c_str(), H5P_DEFAULT);
+	if (exist <= 0) {
+		hid_t temp = H5Gcreate2(m_file, datestr.c_str(), H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+		H5Gclose(temp);
+	}
+	hid_t group = H5Gopen2(m_file, datestr.c_str(), H5P_DEFAULT);
+
+	WriteCluster(clusters.m_households, group, ClusterType::Household);
+	WriteCluster(clusters.m_school_clusters, group, ClusterType::School);
+	WriteCluster(clusters.m_work_clusters, group, ClusterType::Work);
+	WriteCluster(clusters.m_primary_community, group, ClusterType::PrimaryCommunity);
+	WriteCluster(clusters.m_secondary_community, group, ClusterType::SecondaryCommunity);
+
+	H5Gclose(group);
+}
+
+void CheckPoint::WriteCluster(const std::vector<Cluster>& clvector, hid_t& group, const ClusterType& t)
+{
+	std::string dsetname = ToString(t);
+
+	unsigned int totalSize = 0;
+	std::vector<unsigned int> sizes;
+
+	for (auto& cluster : clvector) {
+		totalSize += cluster.GetSize() + 1;
+		sizes.push_back(cluster.GetSize());
+	}
+	hid_t newType = H5Tcreate(H5T_COMPOUND, sizeof(clusterData));
+
+	H5Tinsert(newType, "ID", HOFFSET(clusterData, ID), H5T_NATIVE_UINT);
+	H5Tinsert(newType, "PersonID", HOFFSET(clusterData, PersonID), H5T_NATIVE_UINT);
+
+	hsize_t dims = totalSize;
+	hid_t dataspace = H5Screate_simple(1, &dims, nullptr);
+
+	hid_t dataset =
+	    H5Dcreate2(group, dsetname.c_str(), newType, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+	unsigned int spot = 0;
+	for (auto& cluster : clvector) {
+		auto people = cluster.GetPeople();
+
+		clusterData data[cluster.GetSize() + 1];
```

Comment:
Please don't use a VLA to store these `clusterData` instances. This construct can easily overflow the stack. You can rewrite this as:

```cpp
std::vector<clusterData> data{cluster.GetSize() + 1};
```

and replace `&data` with `data.data()`.